Fermikit

mvdbeek opened the pull request:
* [x] I have read the guidelines above.
* [x] This PR adds a new recipe.
* [ ] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

bgruening commented on recipes/fermikit/build.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+#strictly use anaconda build environment
+CC=${PREFIX}/bin/gcc
```
This should not be needed, imho.

bgruening commented on recipes/fermikit/meta.yaml:
```diff
@@ -0,0 +1,30 @@
+requirements:
+  build:
+  - gcc
+  - zlib
```
I'm wondering if this is needed if you don't need it as runtime-dep

mvdbeek replied:
It should be a runtime dep as well.

druvus commented:
All required (most?) submodules exists as separate bioconda recipes so it would be better to depend on them. As far as I remember fermi-kit is just a perl wrapper for those tools so you would only need/modify `run-calling` and add `k8`.  The pure fermi-kit source code file would be preferred to use [instead](https://github.com/lh3/fermikit/archive/v0.13.tar.gz).

mvdbeek commented:
> All required (most?) submodules exists as separate bioconda recipes so it would be better to depend on them.

I'd argue that the author has put together the exact versions that are known to work well together. I'm worried about using different versions -- what if a user reports that fermikit 0.13 appears to have a bug. Now is that a bug that has been fixed already or not? Difficult to say without further diagnosis if bioconda's fermikit now comes with different versions than the ones it has been released with.
So if that's the route to go I think we'll have to add all of the exact referenced commits.

druvus commented:
I understand that. Maybe you could have all files in ´/opt/fermi-kit´ or similar and just link to the run script from bin folder? I am just worried to have `bwa` and other tools overwritten when I am installing fermi-kit.

mvdbeek commented:
Hmm, that's also a good point. Do you know a recipes that's doing this?

mvdbeek commented:
I think that will be problematic as well .. if you follow the instructions in the README you'd have to use fermi2.pl -- so that would definitely mask fermi2. I think it's best to split things up, that way you know when there is a conflict.

mvdbeek commented:
Alright, I have split up everything, added missing or outdated dependencies and pinned the build versions for fermikit. Does that make sense to you, @druvus @bgruening ?

bgruening commented on recipes/fermikit/test-data/test.fastq:
```diff
@@ -0,0 +1,100 @@
+@phiX174_29_494_0_1_0_0_2:0:0_3:0:0_0
```
We don't include functional tests in BioConda. Can you come up with a easier example that don't need this external file.

mvdbeek replied:
I don't think a reasonable test can be written without a fastq file, you'd probably not detect runtime dependency problems. And bioconda would hugely benefit from this practice. My feeling is that a good fraction of packages here are broken (perl, bioperl, fastx, gnuplot among the ones I recently came across). Functional tests would be a good way to notice this early. I can do the simply help-text grep thing if you prefer, but I'd strongly advocate small functional tests if possible.

bgruening replied:
We can simply not effort functional tests. Space limitations and so on come to my mind. bioperl is known to be broken and there are several issues. But this has nothing to do with tests, this is perl ;) - the perl packages has some good tests. fastx and similar packages were broken because they were created before we started to test packages with busybox. I hope this is fixed gradually from now on.
Small functional tests are ok, we compile libraries if needed, do some tricks like ``echo ">foo\n ACGT > test.fa" and so on, but we should keep this to a absolute minimum until we get more space and more funding.

One thing you could do without populating the github repo is to move this into the build script, fetch the fasta etc ... not ideal, but better than nothing.

mvdbeek replied:
> We can simply not effort functional tests. Space limitations and so on come to my mind.

I agree for the space limitations, but this is just 100 lines here.

> bioperl is known to be broken and there are several issues

it used to work ... 1.6.924=pl5.22.0_0 still works, while newer packages are broken. Of course having tests wouldn't make it easier to upgrade the recipe, but at least you had some level of confidence that installing bioperl will give you a halfway functional package.

> One thing you could do without populating the github repo is to move this into the build script, fetch the fasta etc ... not ideal, but better than nothing.

In principle I like this, but it wouldn't test the runtime dependency. What would be cool is if you could stick urls in the `files:` section.

bgruening replied:
As I said, the perl disaster has different reasons. It's somewhere on my todo list :(

bgruening commented:
@mvdbeek there is something strange happening here.
The tests should fail, simply because our Docker testing in busybox does not support the external file.
It works like this:

* Install fermikit in a build-box
* copy the installation into a runtime-box (busybox)
  * at this time the additional file is not present anymore
* test the command in the runtime box

So this fasta file is not existent in the runtime environment, yet the tests seems to pass. I'm not sure why.

mvdbeek commented:
There are other recipes that use test-data (gnuplot, trf), are you sure this is correct?

bgruening commented:
> There are other recipes that use test-data (gnuplot, trf), are you sure this is correct?

My guess is that the packages were not updated since our last testing updates with busybox.

mvdbeek commented:
Acutally I should be able to use the test-data that comes with fermikit, right?
http://conda.pydata.org/docs/building/meta-yaml.html#source-files

bgruening commented:
Yes.

bgruening commented:
Thanks Marius!

mvdbeek commented:
thanks @druvus @bgruening !
